tests/fXmlRpc/Transport/RecorderTest.php

```diff
@@ -0,0 +1,119 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+namespace fXmlRpc\Transport;
+
+use fXmlRpc\Client;
+
+/**
+ * Class RecorderTest
+ *
+ * @author Piotr Olaszewski <piotroo89 [%] gmail dot com>
+ */
+class RecorderTest extends \PHPUnit_Framework_TestCase
+{
+    /** @var TransportInterface */
+    private $transport;
+
+    /** @var Recorder */
+    private $recorder;
+
+    /** @var Client */
+    private $client;
+
+    private $expectedRequest = '<?xml version="1.0" encoding="UTF-8"?>
+<methodCall>
+    <methodName>TestMethod</methodName>
+    <params>
+        <param><value><string>param1</string></value></param>
+        <param><value><int>2</int></value></param>
+        <param><value><struct><member><name>param3</name><value><boolean>1</boolean></value></member></struct></value></param>
+    </params>
+</methodCall>';
+
+    private $expectedResponse = '<?xml version="1.0"?>
+<methodResponse>
+    <params>
+        <param><value><string>Returned string</string></value></param>
+    </params>
+</methodResponse>';
+
+    public function setUp()
+    {
+        parent::setUp();
+        $this->transport = $this->getMock('fXmlRpc\Transport\TransportInterface');
+        $this->recorder = new Recorder($this->transport);
+        $this->client = new Client('http://foo.com', $this->recorder);
+    }
+
+    public function testReturnLastRequest()
+    {
+        $this->transportOk();
+        $this->client->call('TestMethod', ['param1', 2, ['param3' => true]]);
+
+        $lastRequest = $this->recorder->getLastRequest();
+
+        $this->assertXmlStringEqualsXmlString($this->expectedRequest, $lastRequest);
+    }
+
+    public function testReturnLastResponse()
+    {
+        $this->transportOk();
+        $this->client->call('TestMethod', ['param1', 2, ['param3' => true]]);
+
+        $lastResponse = $this->recorder->getLastResponse();
+
+        $this->assertXmlStringEqualsXmlString($this->expectedResponse, $lastResponse);
+    }
+
+    public function testReturnXmlForRequestAndNullForResponseWhenTransportThrowsException()
+    {
+        try {
+            $this->transportFail();
+            $this->client->call('TestMethod', ['param1', 2, ['param3' => true]]);
+        } catch (\Exception $e) {
+        }
+
+        $lastRequest = $this->recorder->getLastRequest();
+        $lastResponse = $this->recorder->getLastResponse();
+
+        $this->assertNotNull($lastRequest);
+        $this->assertNull($lastResponse);
```

Comment:
There is one behaviour left that is uncovered and potentially confusing:

 - First request: everything works and request and response are populated
 - Second request: exception is thrown, request is populated but response is still the one from the first request which I would say is confusing

In order to fix that case we should cover it and reset the last response before we call the wrapped transport.